### ISSUE TITLE
Rule proposal: no-unnecessary-async

### ISSUE THREAD

**Issue by cjoecker** [sindresorhus/eslint-plugin-unicorn#2487]
### Description

The rule would be beneficial in avoiding async functions that are, in reality, synchronous, and no `async` is necessary.
Unnecessary async function breaks unit tests that use fake timers.

### Fail

```js
async function getLast(arr: number[]) {
	return arr.at(-1);
}
```

### Pass

```js
function getLast(arr: number[]) {
	return arr.at(-1);
}
```

```js
async function getLast() {
	return await getLastFromDb();
}
```

### Proposed rule name

no-unnecessary-async

**Comment by sindresorhus** [comment]
I think it's a good idea in general, but without types, it's really infeasible, and we don't have types available here. It may more appropriate for https://github.com/typescript-eslint/typescript-eslint

**Comment by cjoecker** [comment]
@sindresorhus isn't it possible for the rule to check if there is an `await` or `new Promise` inside the function? If there is none, we can know there is no promise used inside the async function and show the error.

**Comment by sindresorhus** [comment]
That would miss functions that return a promise:

```js
async function getLast(arr: number[]) {
	return thisReturnsAPromise();
}
```

**Comment by fregante** [comment]
Pretty sure this exists and it's disabled in XO.

**Comment by fregante** [comment]
https://eslint.org/docs/latest/rules/require-await
https://typescript-eslint.io/rules/require-await/

**Comment by cjoecker** [comment]
@fregante you are right! Those rules do that, thanks for the hint!
